proposals/0000-unify-natural.rst

```diff
@@ -0,0 +1,59 @@
+Unify ``Nat`` and ``Natural``
+=============================
+
+.. author:: Richard Eisenberg, inspired by Rinat Striungis
+.. date-accepted:: 
+.. ticket-url:: 
+.. implemented:: 
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/364>`_.
+.. contents::
+
+GHC currently uses ``GHC.TypeLits.Nat`` to describe compile-time natural numbers
+but ``Numeric.Natural.Natural`` to describe runtime ones. This proposal unifies
+the two by making ``GHC.TypeLits.Nat`` a synonym for ``Numeric.Natural.Natural``.
+
+Motivation
+----------
+GHC should not have two different types to represent natural numbers. One is
+enough.
+
```

Comment:
I would add something like: 

Existence of separated kind Nat and the impossibility of the promotion of usual Natural data type (and data constructors with Natural fields) looks like an unnecessary and inconvenient complication. It forces us to create redundant types with Nat fields in their constructors (it makes these types uninhabited in terms)only for promotional purposes. This is more cumbersome than the promotion of ordinary data types.

Reply:
Indeed, currently one must go through some hoops to get data types with numeric fields promoted. Consider this data type:

```
data T = MkT [Bool] Natural
```

The `[Bool]` field can be promoted just fine, but the `Natural` field can't be:

```
x = MkT [True, False, True] 42  -- works
type X = MkT [True, False, True] 42   -- doesn't work because of 42
```

So one must introduce extraneous type variables as a workaround:

```
data T n = MkT [Bool] n
x = MkT [True, False, True] (42 :: Natural)
type X = MkT [True, False, True] (42 :: Nat)
```

Here, `n` is instantiated to either `Natural` or `Nat`, depending on whether it's used as a type or as a kind. This issue basically affects any user-defined data type that has a numeric field, and this proposal solves it.

Reply:
Of course, there's also a similar issue with `Symbol` / `Text`, but I'd claim fixing one annoyance at a time is not unreasonable.